lib/block_view_lockups.go

```diff
@@ -0,0 +1,2178 @@
+	txn *MsgDeSoTxn, txHash *BlockHash,
+	blockHeight uint32, blockTimestamp int64,
+	verifySignatures bool) (_totalInput uint64, _totalOutput uint64, _utxoOps []*UtxoOperation, _err error) {
+	var utxoOpsForTxn []*UtxoOperation
+
+	// Validate the starting block height.
+	if blockHeight < bav.Params.ForkHeights.ProofOfStake1StateSetupBlockHeight ||
+		blockHeight < bav.Params.ForkHeights.BalanceModelBlockHeight {
+		return 0, 0, nil,
+			errors.Wrapf(RuleErrorLockupTxnBeforeBlockHeight, "_connectCoinUnlock")
+	}
+
+	// Validate the txn TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeCoinUnlock {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectCoinUnlock: called with bad TxnType %s", txn.TxnMeta.GetTxnType().String(),
+		)
+	}
+
+	// Try connecting the basic transfer without considering transaction metadata.
+	totalInput, totalOutput, utxoOpsForBasicTransfer, err := bav._connectBasicTransfer(txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectCoinUnlock")
+	}
+	utxoOpsForTxn = append(utxoOpsForTxn, utxoOpsForBasicTransfer...)
+
+	// Grab the txn metadata.
+	txMeta := txn.TxnMeta.(*CoinUnlockMetadata)
+
+	// Check for a valid profile public key.
+	var profileEntry *ProfileEntry
+	if len(txMeta.ProfilePublicKey) != btcec.PubKeyBytesLenCompressed {
+		return 0, 0, nil, errors.Wrap(RuleErrorDAOCoinInvalidPubKey,
+			"_connectCoinUnlock")
+	}
+	if !txMeta.ProfilePublicKey.IsZeroPublicKey() {
+		profileEntry = bav.GetProfileEntryForPublicKey(txMeta.ProfilePublicKey.ToBytes())
+		if profileEntry == nil || profileEntry.isDeleted {
+			return 0, 0, nil, errors.Wrap(RuleErrorCoinUnlockOnNonExistentProfile,
+				"_connectCoinUnlock")
+		}
+	}
+
+	// Convert the TransactorPublicKey to HODLerPKID
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockInvalidHODLerPKID,
+			"_connectCoinUnlock")
+	}
+	hodlerPKID := transactorPKIDEntry.PKID
+
+	// Convert the ProfilePublicKey to ProfilePKID.
+	var profilePKID *PKID
+	if txMeta.ProfilePublicKey.IsZeroPublicKey() {
+		profilePKID = ZeroPKID.NewPKID()
+	} else {
+		profilePKIDEntry := bav.GetPKIDForPublicKey(txMeta.ProfilePublicKey.ToBytes())
+		if profilePKIDEntry == nil || profilePKIDEntry.isDeleted {
+			return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockInvalidProfilePKID,
+				"_connectCoinUnlock")
+		}
+		profilePKID = profilePKIDEntry.PKID
+	}
+
+	// Retrieve unlockable locked balance entries.
+	unlockableLockedBalanceEntries, err := bav.GetUnlockableLockedBalanceEntries(
+		hodlerPKID, profilePKID, blockTimestamp)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectCoinUnlock")
+	}
+	if len(unlockableLockedBalanceEntries) == 0 {
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockNoUnlockableCoinsFound,
+			"_connectCoinUnlock")
+	}
+
+	// Unlock coins until the amount specified by the transaction is deducted.
```

Comment:
What is this comment saying? It looks like you unlock coins until you run out of balance entries, not until an amount is deducted.